src/lib/MetamaskDelegatedAccess.ts

```diff
@@ -0,0 +1,82 @@
+import { Wallet } from "ethers"
+import { Buffer } from 'buffer'
+import { encrypt } from "eth-sig-util"
+import { bufferToHex } from 'ethereumjs-util'
+import {StreamrClient} from "streamr-client"
+
+declare global {
+    interface Window { 
+        ethereum: any
+    }
+}
+
+export class MetamaskDelegatedAccess {
+    metamaskAddress?: string 
+    clientAddress?: string
+
+    constructor(){
+    }
+
+    public async connect(): Promise<{client: StreamrClient, address: string}>{
+        // enable and fetch metamask account
+        await window.ethereum.enable()
+        this.metamaskAddress = window.ethereum.selectedAddress
+        
+        // get the delegated key 
+        let sessionWallet: Wallet
+        const encryptedPrivateKey = localStorage.getItem('streamr-chat-encrypted-session-key')
+
+        if (!encryptedPrivateKey){
+            // generate delegated privateKey, encrypt, store and use
+            sessionWallet = await this.createAccount()
+        } else {
+            // decrypt and use
+            sessionWallet = await this.decryptAccount(encryptedPrivateKey)
+        }
+
+        return {
+            client: new StreamrClient({
+                auth: { 
+                    privateKey: sessionWallet.privateKey
+                }
+            }),
+            address: this.metamaskAddress!,
+        }
+    }
+
+    private async createAccount(): Promise<Wallet>{
+        const wallet = Wallet.createRandom()
+        this.clientAddress = wallet.address
+        const encryptionPublicKey = await window.ethereum.request({
+            method: 'eth_getEncryptionPublicKey',
+            params: [this.metamaskAddress]
+        })
+
+        const encryptedMessage = bufferToHex(
```

Comment:
Can you unpack what's going on here?

Reply:
sure thing!

in the brubeck era every message requires signing so we need to create a delegated wallet to connect with the Streamr Client and then secure it using Metamask.

upon first execution (=whenever there's no localStorage item under the `streamr-chat-encrypted-session-key` key) the `createAccount` method gets called. It creates a private key and requests Metamask for the connected account's encryption key. With this two pieces, the created private key is encrypted and stored in localStorage.

whenever the localStorage encrypted key is found the `decryptAccount` method gets called, requesting the connected Metamask account to decrypt the private key.

Reply:
Got it, neat 👍